{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import matplotlib.pyplot as plt\n",
    "import glob\n",
    "import seaborn as sns\n",
    "from scipy.stats import norm, zscore\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "from collections import defaultdict\n",
    "import umap\n",
    "import warnings\n",
    "from scipy.cluster import hierarchy\n",
    "\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "sys.path.append('../funcs')\n",
    "\n",
    "import proteomics as prot\n",
    "import processing as pp\n",
    "import plotting as pl\n",
    "from utils import get_pcs, remove_covariates, plot_pca, plot_pca_gridplot, plot_pca_ax\n",
    "import nmf_utilities as nmu\n",
    "import signatureanalyzer as sa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = \"../data/processed/061721\"\n",
    "NMF_FILE = \"signatures/061721_imputed_res_reg/nmf_output.h5\"\n",
    "PLOTS_DIR = \"figures/ardnmf/061721_imputed_res_reg\"\n",
    "MAPPING_FILE = \"gs://cptac-pancan-getzlab/mapping/PanCan_Participant_freeze_mapping.tsv\"\n",
    "MUT_SIG = \"gs://cptac-pancan-getzlab/Freeze_061721/mutational_signatures/Full_H_compressed_v1.1.tsv\"\n",
    "DENDRO_GROUPS = \"gs://cptac-pancan-getzlab/Freeze_061721/dendrogram/dendro_map_061721.tsv\"\n",
    "H_CORR = \"gs://cptac-pancan-getzlab/Freeze_061721/dendrogram/cosine_sim_H.tsv\"\n",
    "RUN_NAME = NMF_FILE.split(\"/\")[-2]\n",
    "\n",
    "DIFF = 0.1\n",
    "MAX_NORM = 0.25\n",
    "\n",
    "os.makedirs(PLOTS_DIR, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Load Mutational Signatures\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "H_mut = pd.read_csv(MUT_SIG, sep='\\t', index_col=0)\n",
    "mut2Prot_dict = pd.read_csv(MAPPING_FILE,sep='\\t', index_col=0).set_index('WXS').to_dict()['Proteome_Sample_ID']\n",
    "H_mut.index = H_mut.index.map(mut2Prot_dict)\n",
    "H_mut = H_mut.drop('X03BR011')\n",
    "\n",
    "# Thresholds\n",
    "H_mut_apobec = H_mut.apply(lambda x: x['APOBEC'] if ((x.div(x.sum())['APOBEC']>=0.2) and (x['APOBEC']>20)) else 0, 1)\n",
    "H_mut_hrd = H_mut.apply(lambda x: x['HRD'] if x['HRD']>=45 else 0, 1)\n",
    "H_mut_smoking = H_mut.apply(lambda x: x['Tobacco Smoking'] if ((x.div(x.sum())['Tobacco Smoking']>=0.1) and (x['Tobacco Smoking']>10)) else 0, 1)\n",
    "\n",
    "H_mut['HRD'] = H_mut_hrd\n",
    "H_mut['APOBEC'] = H_mut_apobec\n",
    "H_mut['Tobacco Smoking'] = H_mut_smoking\n",
    "H_mut['TMB'] = H_mut.sum(1)\n",
    "\n",
    "HRD_patients = H_mut.index[H_mut['HRD']>0].to_list()\n",
    "\n",
    "pastel_col = sns.color_palette('pastel',10)\n",
    "bright_col = sns.color_palette('bright',10)\n",
    "\n",
    "mut_colors = {\n",
    "    'POLE':\"#5DDCF2\",\n",
    "    'POLE+MSI':\"#1E88E5\",\n",
    "    'CpG>T + Short Indels':\"#9250FE\",\n",
    "    'APOBEC':\"#004D40\",\n",
    "    'CpG>T': \"#DCBEB3\",\n",
    "    'MMRD':\"#FFC107\", \n",
    "    'Unknown (SBS8 0.64)':\"#41B261\",\n",
    "    'Tobacco Smoking':\"#450F6F\",\n",
    "    'UV Exposure': \"#FE98AB\",\n",
    "    'HRD': \"#D81B60\",\n",
    "    'NHR1: Unknown(SBS40 0.94)': \"#CBE07E\"\n",
    "}\n",
    "\n",
    "# Include ESTIMATE Score for plotting\n",
    "resd = prot.loadImmuneDeconv(\"Fig_immuno_metabolism/ImmuneDeconv\")\n",
    "H_mut = H_mut.join(resd['estimate']['ESTIMATEScore'])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. DendroGram\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_raw_mat = pd.read_csv(H_CORR, sep='\\t', index_col=0)\n",
    "C = prot.ss_linkage_cut(corr_raw_mat)\n",
    "groups_df = prot.ss_linkage_groups(C, lim=39)\n",
    "\n",
    "samples_df = pd.read_csv(MAPPING_FILE, sep='\\t', index_col=0)\n",
    "meta = samples_df.set_index('Proteome_Sample_ID')['cohort'].loc[corr_raw_mat.columns]\n",
    "meta.name = 'cohort'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cluster information\n",
    "#clust_df = pd.read_csv('signatures/061721_imputed_res_reg/mappings_pancan_102322.tsv', sep='\\t',index_col=0)\n",
    "clust_df = pd.read_csv('../analysis/signatures/061721_imputed_res_reg/Hierarchical_Clustering/cluster_map.tsv', sep='\\t', index_col=0)\n",
    "clust_df = clust_df.loc[corr_raw_mat.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Expression signatures\n",
    "sigs_order = [\n",
    "    'S20',\n",
    "    'S13',\n",
    "    'S8',\n",
    "    'S5',\n",
    "    'S29',\n",
    "    'S1',\n",
    "    'S27',\n",
    "    'S11',\n",
    "    'S7',\n",
    "    'S12',\n",
    "    'S19',\n",
    "    'S23',\n",
    "    'S18',\n",
    "    'S24',\n",
    "    'S28',\n",
    "    'S6',\n",
    "    'S17',\n",
    "    'S21',\n",
    "    'S16',\n",
    "    'S4',\n",
    "    'S3',\n",
    "    'S2',\n",
    "    'S15',\n",
    "    'S25',\n",
    "    'S30',\n",
    "    'S9',\n",
    "    'S22',\n",
    "    'S14',\n",
    "    'S26',\n",
    "    'S32',\n",
    "    'S10',\n",
    "    'S33',\n",
    "    'S31'\n",
    "]\n",
    "\n",
    "W = pd.read_hdf(NMF_FILE, \"W\")\n",
    "DATA_DIR = \"../data/processed/061721\"\n",
    "pmap_df = pd.read_csv(os.path.join(DATA_DIR, \"var_map_full.tsv\"),sep='\\t',index_col=0)\n",
    "\n",
    "# Count features\n",
    "W = W[['max_id','max','max_norm']]\n",
    "W['n'] = [-1 if x else 1 for x in W.index.str.endswith('_n')]\n",
    "W['max'] = W['max']*W['n']\n",
    "W['max_norm'] = W['max_norm']*W['n']\n",
    "W.index = W.index.str.replace('_n','')\n",
    "W = W.join(pmap_df[['geneSymbol','feature']])\n",
    "W['feature'] = W['feature'].str.capitalize()\n",
    "\n",
    "W_count_df = W.groupby(['max_id','feature']).size().reset_index().pivot(index='max_id',columns='feature',values=0)\n",
    "\n",
    "# Load Hraw\n",
    "Hraw = pd.read_hdf(NMF_FILE, \"Hraw\").T\n",
    "Hraw = Hraw.loc[:,sigs_order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_file = \"../data/processed/061721/rna/tumor_rna_tpm_norm_combat_X.parquet\"\n",
    "protein_file = \"../data/processed/061721_genecentric/imputed/proteome_X.tsv.gz\"\n",
    "phosphoprotein_file = \"../data/processed/061721/imputed_res/phosphoproteome_X.parquet\"\n",
    "\n",
    "var_map_file = \"../data/processed/061721/var_map_full.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cluster information\n",
    "clust_df = pd.read_csv('../analysis/signatures/061721_imputed_res_reg/Hierarchical_Clustering/cluster_map.tsv', sep='\\t', index_col=0)\n",
    "\n",
    "# Raw data\n",
    "rna_df = pd.read_parquet(rna_file).loc[:,clust_df.index]\n",
    "phosph_df = pd.read_parquet(phosphoprotein_file).loc[:,clust_df.index]\n",
    "prot_df = pd.read_csv(protein_file, sep='\\t', index_col=0).loc[:,clust_df.index]\n",
    "\n",
    "# Mapping file\n",
    "map_df = pd.read_csv(var_map_file, sep='\\t', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Union\n",
    "from scipy.stats import zscore\n",
    "\n",
    "def _feature_heatmap(\n",
    "    mat_df: pd.DataFrame,\n",
    "    ax = None,\n",
    "    fig = None,\n",
    "    cbar_ax = None,\n",
    "    dax = None,\n",
    "    metric: str = 'euclidean',\n",
    "    method: str = 'ward',\n",
    "    color_thresh_scale: float = 0,\n",
    "    figsize: tuple = (8,6),\n",
    "    p: int = 30,\n",
    "    vmax: Union[float, None] = None,\n",
    "    vmin: Union[float, None] = None,\n",
    "    cmap: str = 'coolwarm'\n",
    "    ):\n",
    "    \"\"\"\n",
    "    Feature heatmap with Dendrogram.\n",
    "    \"\"\"\n",
    "    from scipy.cluster import hierarchy\n",
    "    import scipy.cluster.hierarchy as shc\n",
    "    from sklearn.cluster import AgglomerativeClustering\n",
    "    import matplotlib as mpl\n",
    "    from matplotlib.pyplot import cm\n",
    "\n",
    "    # Create heatmap\n",
    "    if ax is None and fig is None:\n",
    "        fig, ax = plt.subplots(figsize=figsize)\n",
    "    \n",
    "    if vmax is None:\n",
    "        cbar_top_lim = np.max(mat_df.values)\n",
    "    else:\n",
    "        cbar_top_lim = vmax\n",
    "\n",
    "    if vmin is None:\n",
    "        cbar_bottom_lim = 0\n",
    "    else:\n",
    "        cbar_bottom_lim = vmin\n",
    "    \n",
    "    # Compute feature order\n",
    "    d_linkage = shc.linkage(mat_df, metric=metric, method=method)\n",
    "    dres = shc.dendrogram(d_linkage, p=p, no_plot=True)\n",
    "    dgram_idx = list(map(int, dres['ivl']))\n",
    "\n",
    "    sns.heatmap(\n",
    "        mat_df.iloc[dgram_idx,:].values,\n",
    "        ax=ax,\n",
    "        cbar_ax=cbar_ax,\n",
    "        cbar_kws = {'ticks':[cbar_bottom_lim, cbar_top_lim]},\n",
    "        rasterized=True,\n",
    "        vmax=vmax,\n",
    "        vmin=vmin,\n",
    "        cmap=cmap\n",
    "    )\n",
    "\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "\n",
    "    x0 = ax.get_position().x0\n",
    "    x1 = ax.get_position().x1\n",
    "    y0 = ax.get_position().y0\n",
    "    y1 = ax.get_position().y1\n",
    "\n",
    "    buf = y1*0.015\n",
    "    \n",
    "    if dax is None:\n",
    "        dax = fig.add_axes([x0-0.15-buf, y0, 0.15, y1*0.86])\n",
    "\n",
    "    cmap = cm.rainbow(np.linspace(0, 1, 10))\n",
    "    hierarchy.set_link_color_palette([mpl.colors.rgb2hex(rgb[:3]) for rgb in cmap])\n",
    "\n",
    "    dres = shc.dendrogram(\n",
    "        d_linkage,\n",
    "        p=p,\n",
    "        ax=dax,\n",
    "        above_threshold_color=\"grey\",\n",
    "        color_threshold=color_thresh_scale*max(d_linkage[:,2]),\n",
    "        orientation='left'\n",
    "    )\n",
    "\n",
    "    dax.set_xticks([])\n",
    "    dax.set_yticks([])\n",
    "    [dax.spines[x].set_visible(False) for x in ['top','right','bottom','left']]\n",
    "    \n",
    "def grab_msigdb_genes(gmt, term):\n",
    "    \"\"\"Grab genes corresponding to term.\"\"\"\n",
    "    \n",
    "    with open(gmt) as f:\n",
    "        annots = {}\n",
    "        for line in f:\n",
    "            line_s = line.strip().split('\\t')\n",
    "            annots[line_s[0]] = line_s[2:]\n",
    "    \n",
    "    if isinstance(term, list):\n",
    "        genes = list()\n",
    "        for t in term:\n",
    "            genes += annots[t]\n",
    "        \n",
    "        return list(set(genes))\n",
    "    else:\n",
    "        return annots[term]\n",
    "\n",
    "def filt_feats_matrix(\n",
    "    mat_df, \n",
    "    map_df, \n",
    "    pw=\"HALLMARK_DNA_REPAIR\", \n",
    "    gmt=\"../data/ref/gmts/h.all.v7.0.symbols.gmt.txt\"):\n",
    "    \"\"\"\n",
    "    Grab features from matrix.\n",
    "    \n",
    "    \"\"\"\n",
    "    # Get genes\n",
    "    gene_list = grab_msigdb_genes(gmt, pw)\n",
    "    \n",
    "    # Filter\n",
    "    feature_indices = np.intersect1d(map_df[map_df['geneSymbol'].isin(gene_list)].index, mat_df.index)\n",
    "    \n",
    "    if feature_indices.shape[0]==0:\n",
    "        feature_indices = np.intersect1d(gene_list, mat_df.index)\n",
    "    \n",
    "    return mat_df.loc[feature_indices,:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Load Expression Differential Expression Data\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PROLIF_SET = [\n",
    "    \"HALLMARK_KRAS_SIGNALING_UP\",\n",
    "    \"HALLMARK_MYOGENESIS\",\n",
    "    \"HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION\",\n",
    "    \"HALLMARK_P53_PATHWAY\",\n",
    "    \"HALLMARK_MITOTIC_SPINDLE\",\n",
    "    \"HALLMARK_APICAL_JUNCTION\"\n",
    "]\n",
    "\n",
    "DDR_SET = [\n",
    "    \"HALLMARK_DNA_REPAIR\",\n",
    "    \"HALLMARK_E2F_TARGETS\",\n",
    "    \"HALLMARK_MYC_TARGETS_V1\",\n",
    "    \"HALLMARK_MYC_TARGETS_V2\",\n",
    "    \"HALLMARK_UV_RESPONSE_UP\",\n",
    "    \"HALLMARK_G2M_CHECKPOINT\",\n",
    "]\n",
    "\n",
    "METAB_SET = [\n",
    "    \"HALLMARK_FATTY_ACID_METABOLISM\",\n",
    "    \"HALLMARK_OXIDATIVE_PHOSPHORYLATION\",\n",
    "    \"HALLMARK_GLYCOLYSIS\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "de_df = pd.read_csv(\"diffexp/061721_raw_dendro/full_de_cohort_cov.tsv\", sep=\"\\t\", index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "features_list = list(rna_df.index) + list(prot_df.index) + list(phosph_df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_de_mat_pw(\n",
    "    de_df, \n",
    "    idx, \n",
    "    features_list, \n",
    "    pw=None, \n",
    "    positive=True, \n",
    "    fc_lim=0, \n",
    "    pval_lim=0.1, \n",
    "    gmt=\"../data/ref/gmts/h.all.v7.0.symbols.gmt.txt\",\n",
    "    ntop_rna=None,\n",
    "    ntop_prot=None,\n",
    "    ntop_phosph=None\n",
    "    ):\n",
    "    \"\"\"Filter De Matrix by pathways.\"\"\"\n",
    "    \n",
    "    if pw is not None:\n",
    "        genes = grab_msigdb_genes(gmt, pw)\n",
    "    else:\n",
    "        genes = set(de_df['gene_name'])\n",
    "    \n",
    "    if positive:\n",
    "        df = de_df[\n",
    "            (de_df['id']==idx) & \n",
    "            (de_df['feature'].isin(['transcriptome','proteome','phosphoproteome'])) &\n",
    "            (de_df['adj.P.Val'] <= pval_lim) &\n",
    "            (de_df['logFC'] > fc_lim) &\n",
    "            (de_df['gene_name'].isin(genes)) & \n",
    "            (de_df['index'].isin(features_list))\n",
    "        ]\n",
    "        \n",
    "        if ntop_rna is None:\n",
    "            ntop_rna = df[df['feature']=='transcriptome'].shape[0]\n",
    "        if ntop_prot is None:\n",
    "            ntop_prot = df[df['feature']=='proteome'].shape[0]\n",
    "        if ntop_phosph is None:\n",
    "            ntop_phosph = df[df['feature']=='phosphoproteome'].shape[0]\n",
    "        \n",
    "        return pd.concat([\n",
    "            df[df['feature']=='transcriptome'].sort_values('gsea_rank', ascending=False).iloc[:ntop_rna,:],\n",
    "            df[df['feature']=='proteome'].sort_values('gsea_rank', ascending=False).iloc[:ntop_prot,:],\n",
    "            df[df['feature']=='phosphoproteome'].sort_values('gsea_rank', ascending=False).iloc[:ntop_phosph,:]\n",
    "        ])\n",
    "        \n",
    "    else:\n",
    "        df = de_df[\n",
    "            (de_df['id']==idx) & \n",
    "            (de_df['feature'].isin(['transcriptome','proteome','phosphoproteome'])) &\n",
    "            (de_df['adj.P.Val'] <= pval_lim) &\n",
    "            (de_df['logFC'] < fc_lim) &\n",
    "            (de_df['gene_name'].isin(genes)) &\n",
    "            (de_df['index'].isin(features_list))\n",
    "        ]\n",
    "        \n",
    "        if ntop_rna is None:\n",
    "            ntop_rna = df[df['feature']=='transcriptome'].shape[0]\n",
    "        if ntop_prot is None:\n",
    "            ntop_prot = df[df['feature']=='proteome'].shape[0]\n",
    "        if ntop_phosph is None:\n",
    "            ntop_phosph = df[df['feature']=='phosphoproteome'].shape[0]\n",
    "        \n",
    "        return pd.concat([\n",
    "            df[df['feature']=='transcriptome'].sort_values('gsea_rank').iloc[:ntop_rna,:],\n",
    "            df[df['feature']=='proteome'].sort_values('gsea_rank').iloc[:ntop_prot,:],\n",
    "            df[df['feature']=='phosphoproteome'].sort_values('gsea_rank').iloc[:ntop_phosph,:]\n",
    "        ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "de1_df = filter_de_mat_pw(de_df, 1, features_list, pw=DDR_SET, \n",
    "                          positive=True, ntop_rna=75, ntop_prot=50, ntop_phosph=25)\n",
    "\n",
    "de2_df = filter_de_mat_pw(de_df, 1, features_list, pw=PROLIF_SET, \n",
    "                          positive=False, ntop_rna=75, ntop_prot=50, ntop_phosph=25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ddr = {x:set(grab_msigdb_genes(\"../data/ref/gmts/h.all.v7.0.symbols.gmt.txt\",x)) for x in DDR_SET}\n",
    "prolif = {x:set(grab_msigdb_genes(\"../data/ref/gmts/h.all.v7.0.symbols.gmt.txt\",x)) for x in PROLIF_SET}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx,row in de1_df.iterrows():\n",
    "    pathways_found = []\n",
    "    \n",
    "    for key in ddr.keys():\n",
    "        if row['gene_name'] in ddr[key]:\n",
    "            pathways_found.append(key)\n",
    "            \n",
    "    de1_df.loc[idx,'pathways_found'] = str(pathways_found)\n",
    "    de1_df.loc[idx,'pathways_found_n'] = int(len(pathways_found))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx,row in de2_df.iterrows():\n",
    "    pathways_found = []\n",
    "    \n",
    "    for key in prolif.keys():\n",
    "        if row['gene_name'] in prolif[key]:\n",
    "            pathways_found.append(key)\n",
    "            \n",
    "    de2_df.loc[idx,'pathways_found'] = str(pathways_found)\n",
    "    de2_df.loc[idx,'pathways_found_n'] = int(len(pathways_found))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_sub_df = rna_df.loc[de1_df[de1_df['feature']=='transcriptome']['index']]\n",
    "prot_sub_df = prot_df.loc[de1_df[de1_df['feature']=='proteome']['index']]\n",
    "phosph_sub_df = phosph_df.loc[de1_df[de1_df['feature']=='phosphoproteome']['index']]\n",
    "\n",
    "rna_sub2_df = rna_df.loc[de2_df[de2_df['feature']=='transcriptome']['index']]\n",
    "prot_sub2_df = prot_df.loc[de2_df[de2_df['feature']=='proteome']['index']]\n",
    "phosph_sub2_df = phosph_df.loc[de2_df[de2_df['feature']=='phosphoproteome']['index']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Color scheme\n",
    "cluster_cmap = [item for sublist in [\n",
    "    sns.color_palette(\"Greys\", 2).as_hex()[::-1],\n",
    "    sns.color_palette(\"Purples\", 5).as_hex()[::-1],\n",
    "    sns.color_palette(\"Blues\", 9).as_hex()[::-1],\n",
    "    sns.color_palette(\"Oranges\", 10).as_hex()[::-1]\n",
    "] for item in sublist]\n",
    "\n",
    "cluster_cmap = {'C'+str(idx+1):x for idx,x in enumerate(cluster_cmap)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Full Composite Figure\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import importlib; importlib.reload(pl)\n",
    "\n",
    "f,d,dres,d_linkage = pl.plot_cmatrix_fig2(\n",
    "    corr_raw_mat,\n",
    "    n_clusters=4,\n",
    "    metas=[],\n",
    "    vmax=0.5,\n",
    "    cbar_label='Cosine Similarity',\n",
    "    metric='euclidean',\n",
    "    method='ward',\n",
    "    cmap='coolwarm',\n",
    "    plot_cluster_lines=True,\n",
    "    dendro_meta=meta,\n",
    "    dendro_flag=False,\n",
    "    dendro_cmap=pl.CPTAC_CMAP,\n",
    "    dendro_groups_df=groups_df,\n",
    "    dendro_meta_clust=clust_df['cluster_id'],\n",
    "    color_thresh_scale=0,\n",
    "    figsize=(12,14),\n",
    "    H_mut=H_mut,\n",
    "    mut_colors=mut_colors,\n",
    "    show_xlab=False,\n",
    "    show_upper_triangle=False,\n",
    "    H_sig=Hraw,\n",
    "    W_prop=W,\n",
    "    ax11_plot=filt_feats_matrix(rna_sub_df, map_df, pw=DDR_SET).T.apply(zscore).T,\n",
    "    ax12_plot=filt_feats_matrix(prot_sub_df, map_df, pw=DDR_SET).T.apply(zscore).T,\n",
    "    ax13_plot=filt_feats_matrix(phosph_sub_df, map_df, pw=DDR_SET).T.apply(zscore).T,\n",
    "    ax21_plot=filt_feats_matrix(rna_sub2_df, map_df, pw=PROLIF_SET).T.apply(zscore).T,\n",
    "    ax22_plot=filt_feats_matrix(prot_sub2_df, map_df, pw=PROLIF_SET).T.apply(zscore).T,\n",
    "    ax23_plot=filt_feats_matrix(phosph_sub2_df, map_df, pw=PROLIF_SET).T.apply(zscore).T\n",
    ")\n",
    "\n",
    "plt.savefig(os.path.join(PLOTS_DIR, \"fig2A_v3.pdf\"), dpi=300, bbox_inches='tight')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "broad",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
